[RDY] runtime: pass cmd as List to s:system()

ooora opened the pull request:
:CheckHealth and :Man would fail when the system shell was set tcsh.
Forcing an sh interpreter allows runtime to be system shell independent.

Commit:
runtime: prepend sh -c to shell commands

:CheckHealth and :Man would fail when the system shell was set tcsh.
Forcing an sh interpreter allows runtime to be system shell independent.

jamessan commented:
Instead of forcing the use of more shell (especially in provider.vim, which mostly doesn't use it), why not just remove the use of shell altogether?  Either `jobstart` or `system` with the command specified as a list doesn't use the shell.

I didn't see anything we couldn't do directly.  The most complex one was piping to awk and cut, but the transformations being done there are trivially done in Vim.

Commit:
Merge remote-tracking branch 'neovim/master' into shell-compatibility

Commit:
runtime: pass cmd lists to system() calls

:CheckHealth and :Man would fail when the system shell was set to tcsh.
Passing a cmd list instead of a cmd string to system() provides system
shell independence.

ooora commented:
jamessan, thanks for the feedback. I've refactored the code to remove shell usage in all but one case in man.vim which requires setting a shell variable before the command.

brcolow commented on runtime/autoload/man.vim:
```diff
@@ -80,6 +76,7 @@ function! s:read_page(path) abort
   " Respect $MANWIDTH, or default to window width.
   let cmd  = 'env MANPAGER=cat'.(empty($MANWIDTH) ? ' MANWIDTH='.winwidth(0) : '')
   let cmd .= ' '.s:man_cmd.' '.shellescape(a:path)
+	let cmd = "sh -c '".cmd."'"
```
Misaligned by one space

Commit:
style: fixed spacing

fixed spacing

jamessan commented:
> I've refactored the code to remove shell usage in all but one case in man.vim which requires setting a shell variable before the command.

Which, preferably, would be addressed by allowing one to specify environment variables in a `jobstart()` call.  That's beyond the scope of this PR, though.

Luckily, that's already being handled with the `env` command, so there still isn't a need to use the shell -- `['env', 'MANPAGER=cat', ...]`.

Commit:
man.vim: changed read_page to use cmd list

changed read_page to use cmd list instead of cmd string.

ooora commented:
jamessan, thanks again. I had tried what you suggested before but it didn't work so I thought system() couldn't handle shell variables. I realized after your post that I had put a period between 'MANPAGER=cat' and (empty($MANWIDTH) ? ' MANWIDTH='.winwidth(0) : '') when it should have been changed to a comma.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -112,7 +106,7 @@ endfunction
 function! s:get_path(sect, name) abort
   if empty(a:sect)
-    let path = system(s:man_cmd.' '.s:man_find_arg.' '.shellescape(a:name))
+    let path = system([s:man_cmd, s:man_find_arg, a:name])
```
Now that we're not using `man 2>/dev/null` will man.vim have a problem? cc @nhooyr

nhooyr replied:
It will, see nhooyr/neoman#5

justinmk replied:
@nhooyr Thanks. So @ooora we need to find some other way to make tcsh happy in the cases where man.vim uses `s:man_cmd`.

ooora commented:
@justinmk, it seems as if you can't just add 2>/dev/null to the end of the argument list to redirect stderr output.

line 2: let path = system(['man', s:man_find_arg, a:name, '2>/dev/null'])
>echo system(['man', s:man_find_arg, a:name, '2>/dev/null'])
2>/dev/null: No such file or directory
No manual entry for 2>/dev/null
/usr/share/man/man1/printf.1

If this is correct then this means using jobstart and ignoring stderr via opts.

We can either duplicate and modify s:system() from autoload/health/provider.vim for the purposes of autoload/man.vim or we can generalize s:system() as a simple wrapper for jobstart and use it in both places.

justinmk commented:
> generalize s:system() as a simple wrapper for jobstart and use it in both places.

I guess let's do that for now. It will be temporary, we're going to enhance jobwait/jobstart.

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: use jobstart() instead of system()

stderr must be ignored when calling man. system() doesn’t support this
so switch to jobstart()

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -1,10 +1,6 @@
-else
-  let s:man_cmd = 'man 2>/dev/null'
-endif
+let s:man_cmd = 'man'
```
As @justinmk mentioned earlier, no need for this anymore.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +67,51 @@ function! man#read_page(ref) abort
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) abort
+  if a:event == 'stdout' || a:event == 'stderr'
+    let self.output .= join(a:data, "\n")
```
I haven't tested but I think that adding a newline at the end is unnecessary.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +67,51 @@ function! man#read_page(ref) abort
+  endif
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
```
An error should be printed in this case.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +67,51 @@ function! man#read_page(ref) abort
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    call jobstop(jobid)
+  elseif s:shell_error != 0
```
Shouldn't this be printing an error?

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +67,51 @@ function! man#read_page(ref) abort
+  let jobid = jobstart(a:cmd, opts)
+
+  if jobid < 1
+    let s:shell_error = 1
```
Error should be displayed to the user here.

Commit:
man.vim: throw errors when job fails

If a job fails throw an error.

ooora commented:
@nhooyr, thanks for the feedback. Fixed the not printing errors issue and removed s:man_cmd. As for adding the newline, it is necessary in order to have opts.output match the output given previously by system(). Otherwise path gets chopped off by one character and the output of the man page is printed on a single line.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -67,14 +67,51 @@ function! man#read_page(ref) abort
 
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) abort
+  if a:event == 'stdout' || a:event == 'stderr'
```
we want to ignore stderr for man.vim, right? So that should be removed from the condition

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: remove stderr event from s:system_handler

remove stderr event from s:system_handler since it doesn’t get used.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +65,56 @@ function! man#read_page(ref) abort
+
+" Run a system command and timeout after 30 seconds.
+function! s:system(cmd, ...) abort
+  let stdin = a:0 ? a:1 : ''
```
We can remove the handling of stdin, we don't ever need that.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +65,56 @@ function! man#read_page(ref) abort
+  let jobid = jobstart(a:cmd, opts)
+
+  if jobid < 1
+    throw printf('Command error %d: %s', jobid, 
```
Just for consistency's sake, "Command" should not be capitalized in this error and the ones below.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +65,56 @@ function! man#read_page(ref) abort
+  if res[0] == -1
+    call jobstop(jobid)
+    throw printf('Command timed out: %s',
+        \ type(a:cmd) == type([]) ? join(a:cmd) : a:cmd)
```
No need to handle this either (here, below and above) because we are only going to send a list as the command.

nhooyr commented:
Also, everything that uses s:system should be wrapped in a try catch and call s:error to print errors.

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: added exception handling

catch exceptions and print errors via s:error()

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +65,53 @@ function! man#read_page(ref) abort
+function! s:system_handler(jobid, data, event) abort
+  if a:event == 'stdout'
+    let self.output .= join(a:data, "\n")
+  elseif a:event == 'exit'
```
Could this be simplified to just an `else` because it can only be called on only `stdout` or `exit`?

ooora commented:
@nhooyr, I made the suggested changes. However I have a couple of issues.

1. I had to move the setting of the mandir variable into man#complete() in order to catch and display the error caused by calling s:system(). This means every time man#complete() is called we now need to make two s:system() requests. Is there a better way to handle this?

2. If s:read_page() throws an exception nothing will be put into the man buffer but the UI has already been split. From a UX standpoint it would be nice to only split after you know that something will be written to the buffer. In order to do this s:read_page() would need to be split into s:get_page() and s:put_page().

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -71,16 +65,53 @@ function! man#read_page(ref) abort
+    call jobstop(jobid)
+    throw printf('command timed out: %s', join(a:cmd))
+  elseif s:shell_error != 0
+    throw printf("command error (%d) %s: %s", jobid, join(a:cmd), opts.output)
```
Does this interfere with displaying an error when the manpage is not found?

nhooyr commented:
@ooora 

1. I thought that as well as I was reviewing but `man -w` seems to be very fast so I think it's all good. It's better than before as we were not showing any useful error before. Completion was just silently broken if `man -w` threw an error.
2. 👍. Again, much better than before because we just silently failed if the page could not be read. It's a fair assumption that there would be no error given that we checked the page exists, but even then, it's a good idea not to ignore the error.

nhooyr commented:
For 1, I guess we could lazily set it as a script local variable, but even then, I think that is unnecessary pre optimization.

Commit:
man.vim: split read_page into get_page and put_page

split read_page into get_page and put_page so that window is not split
until we know there is going to be output.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -112,7 +154,7 @@ endfunction
   if empty(a:sect)
-    let path = system(s:man_cmd.' '.s:man_find_arg.' '.shellescape(a:name))
+    let path = s:system(['man', s:man_find_arg, a:name])
     if path !~# '^\/'
```
We shouldn't need this anymore, s:system() will throw an exception because man's error status will be non zero if it does not find a manpage.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -68,19 +66,63 @@ function! man#read_page(ref) abort
+    silent! call jobstop(jobid)
+    throw printf('command timed out: %s', join(a:cmd))
+  elseif s:shell_error != 0
+    throw printf("command error (%d) %s: %s", jobid, join(a:cmd), opts.output)
```
We should start keeping track of stderr because that is where the error will be, not on stdout. Just rename opts.output to opts.stdout and add opts.stderr and record stderr into there.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -68,19 +66,63 @@ function! man#read_page(ref) abort
+  return opts.output
+endfunction
+
+function! s:get_page(path) abort
```
This does not need a try, just call `s:get_page` from within the try's above where it is called.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -68,19 +66,63 @@ function! man#read_page(ref) abort
     return
   endtry
-  call s:read_page(path)
+  let page = s:get_page(path)
```
Once you move s:get_page into the above try, the if empty(page) also becomes unnecessary. Same with the call in man#open_page.

Commit:
man.vim: cleaned up exception handling

removed unnecessary exception handling code. s:system() supports stderr
capture.

ooora commented:
@nhooyr, I made the suggested changes. I've also looked into fixing SunOS support in man.vim. Right now it is completely broken, but I've figured out how to fix it. The hard part is getting neovim to compile on Solaris 11 so I have a test environment.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,70 @@ function! man#open_page(count, count1, mods, ...) abort
+endfunction
+
+function! s:get_page(path) abort
+  return s:system(['env', 'MANPAGER=cat', (empty($MANWIDTH) ? ' MANWIDTH='.winwidth(0) : ''), 'man', a:path])
```
a:path needs shellescape()

nhooyr replied:
There is no need for preceding space in `' MANWIDTH='`, it can be just `'MANWIDTH='`

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -112,18 +148,14 @@ endfunction
-      throw 'no manual entry for '.a:name
-    endif
-    return path
+    return s:system(['man', s:man_find_arg, a:name])
```
a:name needs shellescape()

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -112,18 +148,14 @@ endfunction
   "   - 3pcap section (found on macOS)
   "   - commas between sections (for section priority)
-  return system(s:man_cmd.' '.s:man_find_arg.' -s '.shellescape(a:sect).' '.shellescape(a:name))
+  return s:system(['man', s:man_find_arg, '-s', a:sect, a:name])
```
a:sect and a:name need shellescape()

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,70 @@ function! man#open_page(count, count1, mods, ...) abort
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    silent! call jobstop(jobid)
```
Was `silent!` intended here or did you mean `silent`?

nhooyr commented:
@ooora Sounds good to me but let's get this PR merged first and then you can open another one for Solaris support.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,70 @@ function! man#open_page(count, count1, mods, ...) abort
+  elseif a:event == 'stderr'
+    let self.stderr .= join(a:data, "\n")
+  else
+    let s:shell_error = a:data
```
I find it odd that we have this as a script local variable but the other two as members of self. Perhaps this too should become `self.shell_error`?

ooora replied:
I made the change to use self. I think it makes sense in this context.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,70 @@ function! man#open_page(count, count1, mods, ...) abort
+  let opts = {
+        \ 'stdout': '',
+        \ 'stderr': '',
+        \ 'on_stdout': function('s:system_handler'),
```
Would it be cleaner to have a different function for each event? I'm not too sure, what do you think?

ooora replied:
I would leave it the same function. If we where doing more complex handling then sure.

nhooyr replied:
Alright 👍

ooora commented:
@nhooyr 

1. shellescape() is only needed when you are generating a cmd via str cat'ing. When calling system via a list, you don't need to escape the individual arguments. It's done automatically.

2. Yes, silent! was intentional. I was still getting error output when using just silent.

nhooyr commented:
@ooora yea I realized my mistake and deleted those comments before you responded, sorry! I was hoping you'd never see them 😢

nhooyr commented:
@ooora also, what was the error from jobstop?

ooora commented:
@nhooyr, jobstop will throw errors if jobid isn't a number or if it can't find the job to stop. I figure just ignore the error because we are beyond the point of no return in terms of failing.

nhooyr commented:
@ooora well jobid is a number so it can't return an error there. if It can't find the job to stop, that means we didn't really fail as the command managed to complete so we should just ignore the timeout and return opts.stdout rather throwing an error?

Commit:
vim: make shell_error a member of opts

make shell_error a member of opts.

ooora commented:
I would keep the timeout just for consistency. It's an async function so it is a good to have timeouts. One could imagine the man pages could be on the network via 9p and access may timeout.

nhooyr commented:
@ooora I don't mean we should take out the timeout completely. I mean that when jobwait() times out, before we call jobstop(), the man command could have potentially completed successfully, which will cause jobstop() to throw an error? So if jobstop() throws an error, we should just return opts.stdout instead of throwing an exception.

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: allow opts.stdout return even after jobwait() timing out

Allow s:system() to return even if jobwait() has timed out. Added case
to handle job interruption.

jamessan commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,75 @@ function! man#open_page(count, count1, mods, ...) abort
 
-function! s:read_page(path) abort
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) abort
```
This should be `function! s:system_handler(jobid, data, event) dict abort`.  The `dict` is required as of the partials patch merge.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,75 @@ function! man#open_page(count, count1, mods, ...) abort
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    if jobstop(jobid)
```
An error would still be displayed if you only check the return value of jobstop, instead wrap it in a try catch and if it throws an error, suppress that error and continue.

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: added exception handling around jobstop()

added exception handling around jobstop().

Commit:
man.vim: throw timed out only if job stopped.

throw timed out only if job stopped.

ooora commented:
I messed up the last change. It should only throw timed out if job is stopped (process is still running), otherwise the job finished in time and function should continue.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,76 @@ function! man#open_page(count, count1, mods, ...) abort
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    silent let stopped = jobstop(jobid)
```
this would be cleaner as

```vim
try
  jobstop(jobid)
  throw printf('command timed out: %s', join(a:cmd))
catch
endtry
```

nhooyr replied:
Maybe it could also just be

`catch /E900/` but I'm not sure.

nhooyr replied:
Nevermind, I was wrong, it doesn't work. The original one in my snippet above from `:h :catch` is correct.

Commit:
man.vim: added exception handling around jobstop()

added back exception handling around jobstop()

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,77 @@ function! man#open_page(count, count1, mods, ...) abort
+  elseif res[0] == -2
+    throw printf('command interrupted: %s', join(a:cmd))
+  endif
+  if opts.shell_error != 0
```
We're not using a shell so this should be exit_code.

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
man.vim: change shell_error to exit_code

change shell_error to exit_code

nhooyr reviewed (approved):
man.vim changes look good to me.

@ooora thanks!

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,77 @@ function! man#open_page(count, count1, mods, ...) abort
+function! s:system(cmd, ...) abort
+  let opts = {
+        \ 'stdout': '',
+        \ 'stderr': '',
```
what is this for?

nhooyr replied:
See line 109

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -57,30 +53,77 @@ function! man#open_page(count, count1, mods, ...) abort
+        \ 'stderr': '',
+        \ 'exit_code': 0,
+        \ 'on_stdout': function('s:system_handler'),
+        \ 'on_stderr': function('s:system_handler'),
```
the on_stderr handler isn't needed because we want to ignore stderr results for man.vim

nhooyr replied:
@justinmk we want them so that we can print errors. We ignore them otherwise.

ooora replied:
@justinmk, the reason s:system() is defined the way it is in man.vim is so that health/provider.vim could also use it. I worked on a patch yesterday to do this, but i'm not sure if I should create another PR just for this patch since it is a big refactor. (I had to change the way error reporting was gathered. [return values vs exceptions) The only reason to include the patch in this PR is to fix the fact that I removed a 2>/dev/null from line 419 in health/provider.vim.

justinmk replied:
We want to avoid big refactors and we want to avoid spending lots of time on this. 
We don't care about sharing the function, that's why I mentioned [it will be temporary](https://github.com/neovim/neovim/pull/5734#issuecomment-266091542).

Commit:
Merge branch 'neovim/master' into shell-compatibility

Commit:
provider.vim: remove stderr handling

remove stderr handling

ooora commented:
@justinmk, I just submitted one more commit, but I think i'm done with this PR. If there isn't anymore feedback i'll change the status to ready for merge.

Commit:
Merge branch 'neovim/master' into shell-compatibility

nhooyr reviewed (approved):
provider.vim changes look good too. In the future, we can remove the cruft in `s:systemlist()` as it still handles stdin and some sort of filter argument, but that shouldn't cause any problems for now and we're gonna be changing it anyway so no need to spend time cleaning it up.

Commit:
Merge branch 'neovim/master' into shell-compatibility

justinmk commented:
stderr handling should not be removed for provider.vim. Though one call specifies `>2/dev/null`, other calls need it for providing hints to CheckHealth. I'll add an extra parameter to `s:system()` in provider.vim that ignores stderr optionally.

justinmk commented:
Merged. Thanks @ooora and @nhooyr !
